llvm-hs/src/LLVM/Internal/OrcJIT/LinkingLayer.hs

```diff
@@ -0,0 +1,65 @@
+module LLVM.Internal.OrcJIT.LinkingLayer where
+
+import LLVM.Prelude
+
+import Control.Exception
+import Control.Monad.AnyCont
+import Control.Monad.IO.Class
+import Data.IORef
+import Foreign.Ptr
+
+import LLVM.Internal.Coding
+import LLVM.Internal.ObjectFile
+import LLVM.Internal.OrcJIT
+import qualified LLVM.Internal.FFI.OrcJIT as FFI
+import qualified LLVM.Internal.FFI.PtrHierarchy as FFI
+
+-- | After a 'CompileLayer' has compiled the modules to object code,
+-- it passes the resulting object files to a 'LinkingLayer'.
+class LinkingLayer l where
+  getLinkingLayer :: l -> Ptr FFI.LinkingLayer
+  getCleanups :: l -> IORef [IO ()]
+
+-- | Bare bones implementation of a 'LinkingLayer'.
+data ObjectLinkingLayer = ObjectLinkingLayer {
+   linkingLayer :: !(Ptr FFI.ObjectLinkingLayer),
+   cleanupActions :: !(IORef [IO ()])
```

Comment:
This some name clashes between CompileLayer and this. And modules like `CompileOnDemandLayer`, `IRCompileLayer` which import both `OrcJIT` and `CompileLayer`, would require `hiding` some imports. Also, there's more things to be added here: namely, findSymbol, findSymbolIn. So it might be worth splitting this into a new module now.